lib/mllp/client.ex

```diff
@@ -356,282 +381,383 @@ defmodule MLLP.Client do
+    {:keep_state_and_data, [{:reply, from, format_reply({:error, :busy_with_other_call}, :send)}]}
   end
 
-  defp recv_ack(_state, {time_left, time_owed}, _buffer)
-       when is_integer(time_left) and time_left <= time_owed do
-    {:error, :timeout}
+  def receiving(:state_timeout, :receive_timeout, data) do
+    {:next_state, :connected, reply_to_caller({:error, :timeout}, data)}
   end
 
-  defp recv_ack(state, {time_left, time_owed}, buffer) do
-    {res, elapsed} = do_recv(state, 0, time_left)
+  def receiving(:info, {transport, socket, incoming}, %{socket: socket} = data)
+      when transport in [:tcp, :ssl] do
+    new_data = handle_received(incoming, data)
+    next_state = (new_data.caller && :receiving) || :connected
+    {:next_state, next_state, new_data}
+  end
 
-    case res do
-      {:ok, reply} ->
-        new_buf = buffer <> reply
-        check = byte_size(new_buf) - 3
+  def receiving(:info, {transport_closed, socket}, %{socket: socket} = data)
+      when transport_closed in [:tcp_closed, :ssl_closed] do
+    {:next_state, :disconnected, handle_closed(data)}
+  end
+
+  def receiving(:info, {transport_error, socket, reason}, %{socket: socket} = data)
+      when transport_error in [:tcp_error, :ssl_error] do
+    {:next_state, :disconnected, handle_error(reason, maybe_close(reason, data))}
+  end
 
-        case new_buf do
-          <<@header, _ack::binary-size(check), @trailer>> ->
-            {:ok, new_buf}
+  def receiving(event, unknown, _data) do
+    unexpected_message(:receiving, event, unknown)
+  end
 
-          <<@header, _rest::binary>> ->
-            time_credit = update_recv_time_credit(time_left, time_owed + elapsed)
-            recv_ack(state, time_credit, new_buf)
+  ########################################
+  ### End of :gen_statem callbacks ###
+  ########################################
 
-          _ ->
-            {:error, :invalid_reply}
-        end
+  defp unexpected_message(state, event, message) do
+    Logger.warn(
+      "Event: #{inspect(event)} in state #{state}. Unknown message received => #{inspect(message)}"
+    )
 
-      {:error, _} = err ->
-        err
+    :keep_state_and_data
+  end
+
+  ## Handle the (fragmented) responses to `send` request from a caller
+
+  defp handle_received(_reply, %{caller: nil} = data) do
+    ## No caller, ignore
+    data
+  end
+
+  defp handle_received(<<@header, _rest::binary>> = reply, data) do
+    receive_impl(reply, data)
+  end
+
+  ## No header in the first packet
+  defp handle_received(_reply, %{receive_buffer: []} = data) do
+    reply_to_caller({:error, :invalid_reply}, data)
+  end
+
+  ## The rest of MLLP (after the header was received)
+  defp handle_received(reply, data) do
+    receive_impl(reply, data)
+  end
+
+  def receive_impl(reply, %{receive_buffer: buffer, last_byte_received: last_byte} = data) do
+    new_buf = update_receive_buffer(buffer, reply)
+
+    case trailer_check(reply, last_byte) do
```

Comment:
Cool, we're now dealing with those two edges cases, but I'm not quite sure I understand why we don't just do all that here and (or in one of the handle_received clauses) with some simple pattern matching? Is there something you're seeing that I'm not?

I also realized we're appending vs pre-pending to the buffer, that shouldn't be a huge issue, in theory the acks should be small, but it is non-deterministic 🤔

Reply:
Not sure, it doesn't look much clearer to me than the current implementation. Also, I do not think that pattern matching on the chunk size several times (first handle_receive clause) is much better than using `binary:match/2`.

The current code detects if there is no header in the very first packet, so it doesn't have to do more pattern matching. That makes `header_received=false` to be the same as `receive_buffer = []`, hence we probably don't need an additional field here. 

Wrt prepending: I think [buffer | reply] is a cons-cell operation, so we are not appending here. So it should be O(1) regardless of the ordering. Could of course be wrong here.

Reply:
> Not sure, it doesn't look much clearer to me than the current implementation. Also, I do not think that pattern matching on the chunk size several times is much better than using `binary:match/2`.

Ok cool, I wasn't sure if there was some egregious or something nasty lurking in my suggestion, but it sounds like it's a matter of taste. In that regard, I don't have terribly strong feelings on this one.

> The current code detects if there is no header in the very first packet, so it doesn't have to do more pattern matching. That makes `header_received=false` to be the same as `receive_buffer = []`, hence we probably don't need an additional field here.

This is a great point, you have a bail early clause, mine has to bail late. 



> Wrt prepending: I think [buffer | reply] is a cons-cell operation, so we are not appending here. So it should be O(1) regardless of the ordering. Could of course be wrong here.

Nope! You are 100% right 😄